packages/typeorm-datasource/src/TypeOrmDataSource.ts

```diff
@@ -0,0 +1,321 @@
+
+      if (txnOutputTableExists[0].exists === false) {
+        await ds.query(userDBSchema);
+      }
+
+      const txnOutputIndexExists = await ds.query(txnOutputIndexExistsQuery);
+      if (!txnOutputIndexExists[0].exists) {
+        await ds.query(userDBIndex);
+      }
+    } catch (e) {
+      throw new Error.DBOSError(`Unexpected error initializing schema: ${e}`);
+    } finally {
+      try {
+        await ds.destroy();
+      } catch (e) {}
+    }
+
+    return Promise.resolve();
+  }
+
+  /**
+   * Will be called by DBOS during attempt at clean shutdown (generally in testing scenarios).
+   */
+  async destroy(): Promise<void> {
+    await this.dataSource?.destroy();
+    return Promise.resolve();
+  }
+
+  async checkExecution<R>(
+    client: DataSource,
+    workflowID: string,
+    funcNum: number,
+  ): Promise<
+    | {
+        res: R;
+      }
+    | undefined
+  > {
+    type TxOutputRow = Pick<transaction_completion, 'output'> & {
+      recorded: boolean;
+    };
+
+    const { rows } = await client.query<{ rows: TxOutputRow[] }>(
+      `SELECT output
+          FROM dbos.transaction_completion
+          WHERE workflow_id=$1 AND function_num=$2;`,
+      [workflowID, funcNum],
+    );
+
+    if (rows.length !== 1) {
+      return undefined;
+    }
+    return { res: DBOSJSON.parse(rows[1].output) as R };
+    // return rows[1].output as any;
+  }
+
+  async recordOutput<R>(client: DataSource, workflowID: string, funcNum: number, output: R): Promise<void> {
+    const serialOutput = DBOSJSON.stringify(output);
+    await client.query<{ rows: transaction_completion[] }>(
+      `INSERT INTO dbos.transaction_completion (
+        workflow_id, 
+        function_num,
+        output,
+        created_at
+      ) VALUES ($1, $2, $3, $4)`,
+      [workflowID, funcNum, serialOutput, Date.now()],
+    );
+  }
+
+  /**
+   * Invoke a transaction function
+   */
+  async invokeTransactionFunction<This, Args extends unknown[], Return>(
+    config: TypeOrmTransactionConfig,
+    target: This,
+    func: (this: This, ...args: Args) => Promise<Return>,
+    ...args: Args
+  ): Promise<Return> {
+    const isolationLevel = config.isolationLevel ? config.isolationLevel : 'SERIALIZABLE';
+
+    const readOnly = config?.readOnly ? true : false;
+
+    const wfid = DBOS.workflowID!;
+    const funcnum = DBOS.stepID!;
+    const funcname = func.name;
+
+    // Retry loop if appropriate
+    let retryWaitMillis = 1;
+    const backoffFactor = 1.5;
+    const maxRetryWaitMs = 2000; // Maximum wait 2 seconds.
+    let shouldCheckOutput = false;
+
+    if (this.dataSource === undefined) {
+      throw new Error.DBOSInvalidWorkflowTransitionError('Invalid use of Datasource');
+    }
+
+    while (true) {
+      let failedForRetriableReasons = false;
+
+      if (shouldCheckOutput && !readOnly) {
+        const executionResult = await this.checkExecution<Return>(this.dataSource, wfid, funcnum);
```

Comment:
I think this should be inside the transaction call. I recently updated the other data sources to move check execution inside the transaction